faiss-search/scann/scann.py

```diff
@@ -0,0 +1,94 @@
+            This should be tuned based on recall target
+        :param training_sample_size: The size of the training sample.
+        :param scoring: It can be score_ah (asymmetric hashing) or score_bf (brute force).
+            For small datasets (less than 20k) brute force is recommended
+        :param anisotropic_quantization_threshold: See https://arxiv.org/abs/1908.10396 for
+            a description of this value
+        :param dimensions_per_block: Recommended for AH is 2
+        :param reordering_num_neighbors: Should be higher than the final number of neighbors
+            If this number is increased, the accuracy will increase but it will impact speed
+        """
+        super().__init__(*args, **kwargs)
+        self.num_leaves = num_leaves
+        self.training_iterations = training_iterations
+        self.distance_measure = distance_measure
+        self.num_leaves_to_search = num_leaves_to_search
+        self.training_sample_size = training_sample_size
+        self.scoring = scoring
+        self.anisotropic_quantization_threshold = anisotropic_quantization_threshold
+        self.dimensions_per_block = dimensions_per_block
+        self.reordering_num_neighbors = reordering_num_neighbors
+
+    def build_advanced_index(self, vecs: 'np.ndarray'):
+        """Load vectors into Scann indexers
+        This is a lazy evaluation.
+        The .score_ah(...) and .reorder(...) are creating configuration
+        and only .create_pybind() is building the object
+
+        1) (Optional) The first step is the partitioning, this will be done
+        with .tree(...) during training time,
+        and at query time it will select the top partitions
+        2) The second stage is the Scoring.
+            If partitioning isn't enabled it will measure the distance
+            between the query and all datapoints.
+            If partitioning is enabled it will measure only within the
+            partition to search
+        3) (Optional) This is highly recommended if AH was used.
+        It will take the top k-distances and re-compute the distance.
+        Then the top-k from this new measurement will be selected.
+        """
+        import scann
+
+        searcher = scann.ScannBuilder(vecs, self.training_iterations, self.distance_measure).\
+            tree(self.num_leaves, self.num_leaves_to_search, self.training_sample_size).\
+            score_ah(self.dimensions_per_block, self.anisotropic_quantization_threshold).\
+            reorder(self.reordering_num_neighbors).create_pybind()
+        return searcher
+
+    def query(self, keys: 'np.ndarray', top_k: int, *args, **kwargs) -> Tuple['np.ndarray', 'np.ndarray']:
+        if self.reordering_num_neighbors < self.top_k:
+            self.logger.warning('The number of reordering_num_neighbors should be the same or higher than the number of neighbors')
+        neighbors, distances = searcher.search_batched(keys)
```

Comment:
We need to take care of getting only top-k? Where are we using this param?

Reply:
**We need to take care of getting only top-k?** I think so, the rescore takes only the best distances from scoring and uses only those to rescore
**Where are we using this param?** the neighbors one? this one is then used directly to compute the recall

Reply:
what I mean is that reordering_num_neighbors distances will be returned by `search_batched`. But in the output of `query` we only want `top_k`